### PR TITLE
fetching contributor country when missing confirmed affiliation

### PR DESCRIPTION
**Pull request by truhacevkir**
This part of code describes all changes:
```
var cristinOrganization = fetchCristinOrganizationsById(corporation)
    .or(() -> searchForCristinOrganizationByName(corporation))
    .or(() -> searchForCristinOrganizationByCountry(corporation));
```

When creating contributor from scopus,
And organization contributor is part of does not have ID
And organization contributor is part of is not findable by organization name
Then attempt to set contributor organization to country country organization is from.

### CODE REVIEW

File: expansion/src/test/java/no/unit/nva/expansion/model/cristin/CristinOrganizationTest.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+class CristinOrganizationTest {
+
+    @Test
+    void shouldReturnTrueWhenOrganizationContainsLabelWithValue() {
+        var value = randomString();
+        var cristinOrganization = new CristinOrganization(randomUri(), null, null, List.of(), null,
+                                                          Map.of(randomString(), value));
+
+        assertTrue(cristinOrganization.containsLabelWithValue(value));
```
Comment (torbjokv) [inline expansion/src/test/java/no/unit/nva/expansion/model/cristin/CristinOrganizationTest.java:18]: suggestion: add a test with empty/null value and make sure it returns false

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java [hunk 2]
```diff
@@ -82,39 +75,47 @@ private List<CorporationWithContributors> fillCristinOrganizationData(
...
+        var cristinOrganization = fetchCristinOrganizationsById(corporation)
+            .or(() -> searchForCristinOrganizationByName(corporation))
+            .or(() -> searchForCristinOrganizationByCountry(corporation));
+
+        return cristinOrganization.isPresent()
+                   ? corporation.copy().withCristinCorporations(List.of(cristinOrganization.get())).build()
+                   : corporation;
+    }
+
+    private Optional<CristinOrganization> searchForCristinOrganizationByCountry(
```
Comment (torbjokv) [inline scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:87]: What does this do? get the affiliation country and search for a org with that name? is that correct?

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java [hunk 3]
```diff
@@ -82,39 +75,47 @@ private List<CorporationWithContributors> fillCristinOrganizationData(
...
     private Optional<CristinOrganization> searchCristinOrganization(String organization) {
-        var searchResponse = cristinConnection.searchCristinOrganization(organization);
-        return searchResponse.isPresent() && hasSingleOrganization(searchResponse.get())
-                   ? Optional.ofNullable(getSingleOrganization(searchResponse.get()))
-                   : Optional.empty();
+        return cristinConnection.searchCristinOrganization(organization)
+                   .map(SearchOrganizationResponse::hits)
+                   .orElse(Collections.emptyList())
+                   .stream()
+                   .filter(cristinOrg -> cristinOrg.containsLabelWithValue(organization))
```
Comment (torbjokv) [inline scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:117]: what is the motivation for this filter after the search, maybe add a test that covers that case.

**Review by torbjokv — CHANGES_REQUESTED** [review]
if you can help me understand some of the logic here either in code/tests or just on slack, it's probably good! :-)

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java (hunk 3 above)
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:117]:
```
    @DisplayName("Should create contributor with affiliation created by fetching scopus affiliation group country when could not fetch institution cristin identifier from pia and institution by institution name.")
    @Test
    void shouldCreateContributorAffiliationSearchingAuthorGroupCountryWhenNoResponseFromPiaAndSearchingInstitutionByName() {
        var document = ScopusGenerator.createWithNumberOfContributorsFromAuthorTp(1).getDocument();
        var affiliationCountry = getAffiliationCountry(document);
        var organization = mockSearchOrganizationByNameResponse(affiliationCountry);
        var contributor = contributorExtractorFromDocument(document).generateContributors()
                              .stream()
                              .collect(SingletonCollector.collect());

        var id = ((Organization) contributor.getAffiliations().getFirst()).getId();
        assertThat(id, is(equalTo(organization.id())));
    }
```
When looking up for contributor organization without having organization id, we use cristin-api search organization by name endpoint. This endpoint returns list of organizations, organization name is placed in labels. After we have fetched organizations, we attempt to find the one we need by filtering out organization with label/name which equals label we have from scopus xml file.

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java (hunk 2 above)
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:87]: Almost, countries in Cristin are like organizations. They are returned when you are searching for organization with for example name "Sweden" or "Norway". We want all contributors created under import to have verified affiliation. If we could not fetch organization for contributor, we are trying to fetch contributor country. So when importing this publication, we end up with contributor that has country as affiliation:

image

**Comment by truhacevkir** [comment]
> if you can help me understand some of the logic here either in code/tests or just on slack, it's probably good! :-)

There are multiple ways of extracting contributors from scopus xml. In this case we are extracting contributors from author group, this is a group of authors that has the same affiliation. For example:

```
          <author-group>
            <author auid="7006535102" seq="1" type="auth">
              <ce:initials>K.G.</ce:initials>
              <ce:indexed-name>Foote K.G.</ce:indexed-name>
              <ce:surname>Foote</ce:surname>
              <ce:given-name>Kenneth G.</ce:given-name>
              <preferred-name>
                <ce:initials>K.G.</ce:initials>
                <ce:indexed-name>Foote K.</ce:indexed-name>
                <ce:surname>Foote</ce:surname>
                <ce:given-name>Kenneth G.</ce:given-name>
              </preferred-name>
            </author>
            <affiliation afid="60029622" dptid="112646978" country="nor">
              <organization>Department of Applied Mathematics</organization>
              <organization>University of Bergen</organization>
              <organization>5014</organization>
              <city-group>Bergen</city-group>
              <affiliation-id afid="60029622" dptid="112646978"/>
              <country>Norway</country>
            </affiliation>
          </author-group>
```
In the case where `author auid` is not resolvable (we are not able to fetch person data from Cristin), we use data lying in `affiliation` to create contributor affilation.
We are first trying to find this organization in Cristin - `<organization>Department of Applied Mathematics</organization>`, if no organization like this, we are trying to find `<organization>University of Bergen</organization>` and if there are no hits as well, at the end we are attempting to find `<country>Norway</country>`. So we are setting contributor affiliation to `Norway`.

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java (hunk 2 above)
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:87]: We are not really interested in non Norwegian contributor organizations, we need only to know if there are some international contributors on publication (important for getting more nvi points)

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java (hunk 3 above)
Comment (monaullah) [inline scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:117]: This filter will also affect `searchForCristinOrganizationByName`, should the containsLabelWithValue()-method ignore case when comparing strings?

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/model/CorporationWithContributors.java [hunk 4]
```diff
@@ -102,7 +103,7 @@ private List<Corporation> generateCorporationFromAuthorGroupTp() {
             organizationName -> Map.of(guessTheLanguageOfTheInputStringAsIso6391Code(organizationName),
                                        organizationName))
                          : extractCountryNameAsAffiliation();
-        return isNotNorway(labels.orElse(Map.of()))
+        return isNotNorway(labels.orElse(Map.of())) && name.isPresent() && !name.get().isBlank()
                    ? List.of(new UnconfirmedOrganization(name.orElse(null)))
```
[the remark this answers is not in the document]
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/model/CorporationWithContributors.java:107]: We have to handle nullpointer exception, or I misunderstood?

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java (hunk 3 above)
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:117]: It is the same for organization, we can fetch multiple items from Cristin when for example searching for NTNU, like NTNU, National Taiwan Normal University and NTNU Samfunnsforskning AS. We filtering out organizations that does not match organization name.

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/model/CorporationWithContributors.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (monaullah) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/model/CorporationWithContributors.java:107]: Since you added the `name.isPresent()`-check, you don't need the `.orElse()` method?
```
isNotNorway(labels.orElse(Map.of())) && name.isPresent() && !name.get().isBlank()
                   ? List.of(new UnconfirmedOrganization(name.get()))
                   : List.of();
```
Also, `name.isPresent() && !name.get().isEmpty()` is repeated twice, extract to method?

[the remark this answers is not in the document]
Comment (truhacevkir) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/model/CorporationWithContributors.java:107]: Now I get it, thank you :)

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java [hunk 5]
```diff
@@ -82,39 +75,47 @@ private List<CorporationWithContributors> fillCristinOrganizationData(
     }
 
     private CorporationWithContributors addCristinOrganisationData(CorporationWithContributors corporation) {
-        return nonNull(corporation.getCristinOrganizationId())
-                   ? fetCristinOrganizationsById(corporation)
-                   : searchForCristinOrganizationByName(corporation);
+        var cristinOrganization = fetchCristinOrganizationsById(corporation)
+            .or(() -> searchForCristinOrganizationByName(corporation))
+            .or(() -> searchForCristinOrganizationByCountry(corporation));
```
[the remark this answers is not in the document]
Comment (brinxmat) [reply scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:80]: "Norge AS"? ;)

File: scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java [hunk 6]
```diff
@@ -82,39 +75,49 @@ private List<CorporationWithContributors> fillCristinOrganizationData(
...
+
+        return cristinOrganization.isPresent()
+                   ? corporation.copy().withCristinCorporations(List.of(cristinOrganization.get())).build()
+                   : corporation;
+    }
+
+    //Countries are stored as organizations in Cristin, we use the same endpoint
+    //for fetching countries as we do with organizations
+    private Optional<CristinOrganization> searchForCristinOrganizationByCountry(
```
Comment (brinxmat) [inline scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java:89]: Maybe call this method `searchForCountry` since the organisation in question is a country…I'd also take a moment to consider what an Organization (big-O) is in the ontology "Administrative grouping of people working towards a shared goal", which seems at odds with the current use as I don't think any of the current organizations can be said to be "working towards a shared goal", but "Administrative grouping of people" covers all current cases. Update the ontology accordingly pls.

**Review by brinxmat — CHANGES_REQUESTED** [review]

**Review by brinxmat — APPROVED** [review]
